From a57597026308495e43de1c4ef7e2af2726b60f40 Mon Sep 17 00:00:00 2001
From: Jakub Jirutka <jakub@jirutka.cz>
Date: Tue, 20 Sep 2016 00:40:00 +0200
Subject: [PATCH 06/25] Replace Sys::Filesystem with non-FFI code

Gem sys-filesystem uses FFI which Grsecurity/PaX doesn't like.
It's used only in system_info_controller, so better to replace this
code than enabling MPROTECT on ruby binary.

https://github.com/djberg96/sys-filesystem/issues/17
---
 Gemfile                                         |  3 ++-
 app/controllers/admin/system_info_controller.rb | 35 ++++++++++++++++++++++---
 2 files changed, 34 insertions(+), 4 deletions(-)

diff --git a/Gemfile b/Gemfile
index 4b4c7da..b31736c 100644
--- a/Gemfile
+++ b/Gemfile
@@ -390,7 +390,8 @@ gem 'health_check', '~> 2.6.0'
 
 # System information
 gem 'vmstat', '~> 2.3.0'
-gem 'sys-filesystem', '~> 1.1.6'
+# Uses Ruby FFI which Grsecurity/PaX doesn't like.
+gem 'sys-filesystem', '~> 1.1.6', group: :exclude
 
 # SSH host key support
 gem 'net-ssh', '~> 4.1.0'
diff --git a/app/controllers/admin/system_info_controller.rb b/app/controllers/admin/system_info_controller.rb
index 9903972..0456db1 100644
--- a/app/controllers/admin/system_info_controller.rb
+++ b/app/controllers/admin/system_info_controller.rb
@@ -1,3 +1,5 @@
+require 'open3'
+
 class Admin::SystemInfoController < Admin::ApplicationController
   EXCLUDED_MOUNT_OPTIONS = [
     'nobrowse',
@@ -29,10 +31,14 @@ class Admin::SystemInfoController < Admin::ApplicationController
     'vfat'
   ].freeze
 
+  MOUNT_REGEX = /(\S+) on (\S+) type (\S+) \(([^)]+)\)/
+
+  Mount = Struct.new('Mount', :name, :mount_point, :mount_type, :options)
+  FsStat = Struct.new('FsStats', :path, :bytes_total, :bytes_used)
+
   def show
     @cpus = Vmstat.cpu rescue nil
     @memory = Vmstat.memory rescue nil
-    mounts = Sys::Filesystem.mounts
 
     @disks = []
     mounts.each do |mount|
@@ -42,15 +48,38 @@ class Admin::SystemInfoController < Admin::ApplicationController
       next if (EXCLUDED_MOUNT_TYPES & [mount.mount_type]).any?
 
       begin
-        disk = Sys::Filesystem.stat(mount.mount_point)
+        disk = fs_stat(mount.mount_point)
         @disks.push({
           bytes_total: disk.bytes_total,
           bytes_used:  disk.bytes_used,
           disk_name:   mount.name,
           mount_path:  disk.path
         })
-      rescue Sys::Filesystem::Error
+      rescue IOError
       end
     end
   end
+
+  def mounts
+    stdout, stderr, status = Open3.capture3('mount')
+    fail IOError, stderr unless status.success?
+  
+    stdout.lines
+      .map { |line| MOUNT_REGEX.match(line) }
+      .compact
+      .map { |match| Mount.new(*match.captures) }
+  end
+  
+  def fs_stat(mount_point)
+    stdout, status = Open3.capture2('stat', '-c', '%s %b %a', '-f', mount_point)
+    fail IOError unless status.success?
+  
+    block_size, blocks_total, blocks_free = stdout.split(' ').map(&:to_i)
+  
+    bytes_total = blocks_total * block_size
+    bytes_free = blocks_free * block_size
+    bytes_used = bytes_total - bytes_free
+  
+    FsStat.new(mount_point, bytes_total, bytes_used)
+  end
 end
-- 
2.10.1 (Apple Git-78)

